### DISCUSSION TITLE
Data copying in AM fragments receive handlers

### DISCUSSION THREAD

**Discussion by huzhijiang** [openucx/ucx#9803] [Q&A] [answered]
Hi,
According to https://github.com/openucx/ucx/blob/f3f911f36dd1deeba1d668704a1491c672be1dab/src/ucp/core/ucp_am.c#L1491
it seems if we send AM data in fragments, the receiver side needs to allocate a contig buffer to hold all of them. This could be a potential perfomance bottleneck. Sending data in a single piece will not leading to this copying on the receiver side but sender side multi-rail function will not work either(see https://github.com/openucx/ucx/blob/f3f911f36dd1deeba1d668704a1491c672be1dab/src/ucp/proto/proto_am.inl#L423).

Do we have any other optimal way (can be AM or TAG, HW offload...) to not only avoid fragments copy on receiver side , but also enable multi-rail on sender side?

**Comment by evgeny-leksikov** [comment, marked as answer]
@huzhijiang rndv protocols should enable multi-rail by default and utilize zcopy rdma capability, you can enable it by env var `UCX_RNDV_THRESH=<msg_size>` or control per message by `UCP_AM_SEND_FLAG_RNDV` from your code.

**Reply by huzhijiang** [reply]
Hi @evgeny-leksikov ,
Thanks for the info.
